### PR TITLE
LogsDB data migration integration testing

### LINKED ISSUE #109803: LogsDB - Data Migration Testing
### Description

As part of the effort to test integration of LogsDB we would like to run data-migration testing. We would like to make sure the migration process of log data from existing storage solutions to the new “logsdb” index mode and vice versa works seamlessly. Users should be able to restore indices from snapshots and source-only snapshots and eventually rely on reindexing to migrate existing data.

- [ ] Reindex a "logsdb" index from an index whose index mode is “standard”. Make sure reindexing is successful.
- [ ] Create a snapshot out of an index using index mode is set to “logsdb”. Make sure snapshot creation is successful.
- [ ] Restore a snapshot that was created out of an index whose index mode was set to “logsdb” at snapshot creation time. Make sure restoring the snapshot is successful.
- [ ] Create a source-only snapshot from an index whose index mode is “logsdb”. Make sure snapshot creation is successful.
- [ ] Restore a source-only snapshot that was created out of an index whose index mode was set to “logsdb” at snapshot creation time. We expect an error here since `_source` is missing.

### PR DESCRIPTION
**Pull request by salvatore-campagna**
Here we test reindexing logsdb indices, creating and restoring snapshots. Note that logsdb uses synthetic source and restoring source-only snapshots fails due to missing _source.

Resolves #109803

### CODE REVIEW

File: modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/LogsDataStreamRestIT.java [hunk 1]
```diff
@@ -323,11 +361,199 @@ public void testLogsTimeSeriesIndexModeSwitch() throws IOException {
...
+        deleteIndex(index);
+        restoreSnapshot(repository, snapshot, true);
+
+        final String restoreIndex = randomAlphaOfLength(7).toLowerCase(Locale.ROOT);
+        final Request mountRequest = new Request("POST", "/_snapshot/" + repository + '/' + snapshot + "/_mount");
+        mountRequest.addParameter("wait_for_completion", "true");
+        mountRequest.setJsonEntity("{\"index\": \"" + index + "\",\"renamed_index\": \"" + restoreIndex + "\"}");
+        assertOK(client.performRequest(mountRequest));
+
+        assertThat(
```
Comment (lkts) [inline modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/LogsDataStreamRestIT.java:500]: Same - can we check documents?

**Comment by salvatore-campagna** [comment]
@lkts this was not ready for review yet...

File: modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/LogsDataStreamRestIT.java [hunk 2]
```diff
@@ -323,11 +361,199 @@ public void testLogsTimeSeriesIndexModeSwitch() throws IOException {
...
+                "dest": {
+                  "index": "standard-apache-kafka",
+                  "op_type": "create"
+                }
+            }
+            """);
+        assertOK(client.performRequest(reindexRequest));
+        assertDataStreamBackingIndexMode("standard", 0, "standard-apache-kafka");
+        assertThat(
+            entityAsMap(client.performRequest(new Request("POST", "/standard-apache-kafka/_count"))).get("count"),
```
[the remark this answers is not in the document]
Comment (salvatore-campagna) [reply modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/LogsDataStreamRestIT.java:411]: What do you exactly mean when saying "look correct"?

**Comment by lkts** [comment]
@salvatore-campagna lol sorry about that. Can you mark it as draft then?

**Comment by salvatore-campagna** [comment]
> @salvatore-campagna lol sorry about that. Can you mark it as draft then?

I usually wait at least for a round of CI to be green or, at least to see if there is anything wrong and then I ask for a review selecting reviewers. I do this to avoid wsting people time reviewing something that is not ready yet. I rarely use draft PRs.

File: modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/LogsDataStreamRestIT.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (lkts) [reply modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/LogsDataStreamRestIT.java]: I agree that this is not an ideal place to test reindex operation itself but AFAIK there are currently no tests for this. Do we have an item for that? I was thinking that such a simple document should be identical even with synthetic _source and therefore we could get an initial test for that behavior.

**Review by lkts — APPROVED** [review]

**Review by martijnvg — APPROVED** [review]
lgtm

**Comment by salvatore-campagna** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[](https://github.com/elastic/elasticsearch/pull/113448)|


### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

**Pull request merged**